sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -67,65 +67,537 @@ __ESIMD_API SurfaceIndex get_surface_index(AccessorTy acc) {
+/// @param p is the base pointer.
+/// @param offsets is the zero-based offsets in bytes.
+/// @param pred is predicates.
+/// @param pass_thru contains the vector which elements are copied
+/// to the returned result when the corresponding element of \p pred is 0.
+/// @return is a vector of type T and size N * NElts
+///
+template <typename T, int NElts, lsc_data_size DS, cache_hint L1H,
+          cache_hint L2H, int N, typename OffsetT>
+__ESIMD_API simd<T, N * NElts> gather_impl(const T *p, simd<OffsetT, N> offsets,
+                                           simd_mask<N> pred,
+                                           simd<T, N * NElts> pass_thru) {
+  static_assert(std::is_integral_v<OffsetT>, "Unsupported offset type");
+  check_lsc_vector_size<NElts>();
+  check_lsc_data_size<T, DS>();
+  check_cache_hint<cache_action::load, L1H, L2H>();
+  constexpr uint16_t AddressScale = 1;
+  constexpr int ImmOffset = 0;
+  constexpr lsc_data_size EDS = expand_data_size(finalize_data_size<T, DS>());
+  constexpr lsc_vector_size VS = to_lsc_vector_size<NElts>();
+  constexpr auto Transposed = lsc_data_order::nontranspose;
+  using MsgT = typename lsc_expand_type<T>::type;
+  simd<uintptr_t, N> Addrs = reinterpret_cast<uintptr_t>(p);
+  Addrs += convert<uintptr_t>(offsets);
+  simd<MsgT, N * NElts> PassThruExpanded = lsc_format_input<MsgT>(pass_thru);
+  simd<MsgT, N * NElts> Result =
+      __esimd_lsc_load_merge_stateless<MsgT, L1H, L2H, AddressScale, ImmOffset,
+                                       EDS, VS, Transposed, N>(
+          pred.data(), Addrs.data(), PassThruExpanded.data());
+  return lsc_format_ret<T>(Result);
+}
+
+// Returns true iff it is Ok to use llvm.masked.gather and llvm.masked.scatter.
+// By default (without use specifying __ESIMD_GATHER_SCATTER_LLVM_IR) it is
+// not used because of an issue in GPU driver, which does not recognize
+// those operations in SPIR-V when they are used in mixed (scalar and vector)
+// kernels using invoke_simd() API.
+constexpr bool isMaskedGatherScatterLLVMAvailable() {
+#ifdef __ESIMD_GATHER_SCATTER_LLVM_IR
+  return true;
+#else
+  return false;
+#endif
+}
+
+} // namespace detail
+
+/// template <typename T, int N, int VS, typename OffsetT,
+///           typename PropertyListT = empty_properties_t>
+/// simd<T, N> gather(const T *p, simd<OffsetT, N / VS> byte_offsets,
+///                   simd_mask<N / VS> mask, simd<T, N> pass_thru,
+///                   PropertyListT props = {});                   // (usm-ga-1)
+/// simd<T, N> gather(const T *p, simd<OffsetT, N / VS> byte_offsets,
+///                   simd_mask<N / VS> mask,
+///                   PropertyListT props = {});                   // (usm-ga-2)
+/// simd<T, N> gather(const T *p, simd<OffsetT, N / VS> byte_offsets,
+///                   PropertyListT props = {});                   // (usm-ga-3)
+///
+/// The next 3 functions are similar to the above and added for convenience.
```

Comment:
```suggestion
/// The next 3 functions are similar to the above and were added for convenience.
```

Reply:
Fixed in f541a69d936f5c31befd154497f8f1e9f1af7b57